### PR TITLE
Checking in ProcRecordProducer

### PR DESCRIPTION
**Pull request by ghost**
1. Removing all of the following classes as they are not relevant based on our latest design discussions:
   com/mapr/distiller/server/coordinators/IostatProcessRecordProducer.java
   com/mapr/distiller/server/coordinators/IostatRecordToComponentRecordProducer.java
   com/mapr/distiller/server/coordinators/SystemCpuRecordProducer.java
   com/mapr/distiller/server/coordinators/SystemCpuRecordToSystemCpuHighRecordProducer.java
   com/mapr/distiller/server/coordinators/SystemCpuRecordToSystemCpuMovingAverageRecordProducer.java
   com/mapr/distiller/server/processors/IostatRecordToSplitterRecordProcessor.java
   com/mapr/distiller/server/processors/IostatStderrRecordProcessor.java
   com/mapr/distiller/server/processors/IostatStdoutRecordProcessor.java
   com/mapr/distiller/server/processors/SystemCpuHighRecordProcessor.java
   com/mapr/distiller/server/processors/SystemCpuMovingAverageRecordProcessor.java
   com/mapr/distiller/server/queues/SplitterRecordQueue.java
   com/mapr/distiller/server/readers/PassthroughRecordReader.java
   com/mapr/distiller/server/readers/WholeLineRecordReader.java
   com/mapr/distiller/server/recordtypes/CpuUtilizationSample.java
   com/mapr/distiller/server/recordtypes/DeviceUtilizationSample.java
   com/mapr/distiller/server/recordtypes/IoStatRecord.java
   com/mapr/distiller/server/recordtypes/SplitterRecord.java
   com/mapr/distiller/server/recordtypes/SystemCpuHighRecord.java
   com/mapr/distiller/server/recordtypes/SystemCpuMovingAverageRecord.java
   com/mapr/distiller/server/recordtypes/WholeLineRecord.java
2. Adding all of the following classes that are used to implement ProcRecordProducer.  All of the Record classes are written according to the format set in Record.java from the repo.  Also, I'm listing SystemCpuRecord here as it's a completely different implementation than the previous one that exists in the repo as of now.  The previous SystemCpuRecord is not going to be used.

com/mapr/distiller/server/producers/raw/ProcRecordProducer.java
com/mapr/distiller/server/recordtypes/DiskstatRecord.java
com/mapr/distiller/server/recordtypes/NetworkInterfaceRecord.java
com/mapr/distiller/server/recordtypes/ProcessResourceRecord.java
com/mapr/distiller/server/recordtypes/SystemCpuRecord.java
com/mapr/distiller/server/recordtypes/SystemMemoryRecord.java
com/mapr/distiller/server/recordtypes/TcpConnectionStatRecord.java
com/mapr/distiller/server/recordtypes/ThreadResourceRecord.java
com/mapr/distiller/server/scheduler/GatherMetricEvent.java
com/mapr/distiller/server/scheduler/MetricEventComparator.java
1. I'm not modifying SystemMonitor.java, but if you wan to test out the ProcRecordProducer, you can make this change.  Right before where it says "Monitors are started, now wait until its time to shut down" in the code, add the following lines:

ConcurrentHashMap<String, Integer> rawMetricList = new ConcurrentHashMap<String, Integer>(1000);
rawMetricList.put("cpu.system", 3000);
rawMetricList.put("disk.system",  3000);
rawMetricList.put("memory.system",  3000);
rawMetricList.put("network.interfaces",  3000);
rawMetricList.put("process.resources",  3000);
rawMetricList.put("thread.resources", 3000);
rawMetricList.put("tcp.connection.stats", 3000);
ProcRecordProducer procRecordProducer = new ProcRecordProducer(nameToRecordQueueMap, rawMetricList);
procRecordProducer.start();

This will start a ProcRecordProducer thread that gathers all 7 types of metrics it knows how to gather, each one gathered once every 3 seconds (3000 milliseconds).
1. I'm updating the following existing files:

Record.java - I've added new constructors that get used by ProcRecordProducer record types, removed the durationms long value and associated set method, and calculate it on the fly by difference between timestamp and previousTimestamp.
SystemCpuRecordProcessor.java - I've updated the threshold and average functions to use the format set in the new SystemCpuRecord.java implementation
MovingAverageable.java - Update so the methods throw Exceptions.  This is necessary because a request can be made to average, for instance, two records with the same timestamp, which is not useful/meaningful.  In such a situation, an Exception needs to be thrown so the caller of the average method can handle it however necessary.

### CODE REVIEW

**Comment by vsowrirajan** [comment]
1. Adding libraries and config for XmlRpcServer
   I don't think we need this. As I have already added them as part of pom.xml. So these dependencies will get pulled automatically. And also even if we need some other dependency, lets make it part of maven pom.xml rather we manage it ourselves.

File: Distiller-server/src/main/java/com/mapr/distiller/server/producers/raw/ProcRecordProducer.java [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+	private boolean shouldExit=false;
+	
+	//This holds the list of metrics to gather sorted by the time at which they should be gathered.
+	private TreeSet<GatherMetricEvent> metricSchedule = new TreeSet<GatherMetricEvent>(new MetricEventComparator());
+	
+	//The number of clock ticks (jiffies) per second.  Typically 100 but custom kernels can be compiled otherwise
+	private int clockTick;
+	
+	//The length of time, in milliseconds, represented by a single clock tick (jiffy). Typically 10ms
+	private Double clockTickms;
```
Comment (vsowrirajan) [inline Distiller-server/src/main/java/com/mapr/distiller/server/producers/raw/ProcRecordProducer.java:39]: These "Double" can be "double" primitive types. Unless, we are using it as part of a List or Map we don't need it to be of wrapper type.

File: Distiller-server/src/main/java/com/mapr/distiller/server/recordtypes/ProcessResourceRecord.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+	public String toString(){
+		return super.toString() + " process.resources: " + pid + " " + comm + " " + state + " " + ppid + " " + num_threads;
+	}
+
+}
+/**
+
+	public static ProcessResourceRecord diff(ProcessResourceRecord oldRecord, ProcessResourceRecord newRecord){
+		//This function should be called to diff ProcessResourceRecords when the clockTick field of those records is populated.
+		//If this is called against records without clockTick populated then no cpu utilization will be calculated (because we can't).
```
Comment (vsowrirajan) [inline Distiller-server/src/main/java/com/mapr/distiller/server/recordtypes/ProcessResourceRecord.java:111]: Do we need to keep these commented blocks of code?

File: Distiller-server/src/main/java/com/mapr/distiller/server/recordtypes/SystemCpuRecord.java [hunk 3]
```diff
@@ -1,58 +1,245 @@
...
+	public double getIowaitCpuUtilPct(){
+		return iowaitCpuUtilPct;
+	}
+}
 
-	public void setIdle(double idle) {
-		this.idle = idle;
+	/**
+	public static String[] type1DiffSupport(){
+		return new String[] {	"SystemCpuRecord.idleCpuUtilPct",
```
Comment (vsowrirajan) [inline Distiller-server/src/main/java/com/mapr/distiller/server/recordtypes/SystemCpuRecord.java:182]: Again same, do we need to keep this commented blocks of code?

**Comment by vsowrirajan** [comment]
There is a lot of code to review. I think it will take some time to review every class.  You know more about this code than me. If you think, we can go ahead and merge, I will merge them.

File: Distiller-server/src/main/java/com/mapr/distiller/server/recordtypes/ProcessResourceRecord.java (hunk 2 above)
Comment (ghost) [inline Distiller-server/src/main/java/com/mapr/distiller/server/recordtypes/ProcessResourceRecord.java:111]: Not really.

**Comment by ghost** [comment]
I think we can merge it then.  You're welcome to review asynchronously and
make changes after.  Theres still stuff to cleanup/adjust in this code but
it is functional as is and I was planning to do the cleanup later.

File: Distiller-server/src/main/java/com/mapr/distiller/server/producers/raw/ProcRecordProducer.java (hunk 1 above)
Comment (ghost) [inline Distiller-server/src/main/java/com/mapr/distiller/server/producers/raw/ProcRecordProducer.java:39]: I'll update this for next check in.

**Comment by ghost** [comment]
I put checked that stuff in because I can point Eclipse directly at the
local git repo dir on my dev machine and it is able to compile/run without
me needing to link it into other stuff.  I agree we should remove this
before we package anything we are going to distribute, I wanted to have it
in there for now for convenience.

**Comment by vsowrirajan** [comment]
I am merging this pull request now. 
1. But I think importing Distiller as a Maven project will make it even more convenient for you. 
2. I will also review other source code and slowly we can refine them over the course of time.

**Pull request merged**